File: core_logic/python_package/plantbreeding_py/generate_stripplot_plan.py
import pandas as pd
import random
import matplotlib.pyplot as plt
import matplotlib.patches as patches

def generate_stripplot_plan(wp_factor, sp_factor, n_locs=2, n_blocks=3):
    """
    Generate Strip-Plot Design Randomization Plan
    
    Parameters:
    wp_factor (list): Whole-plot treatment names (e.g., Irrigation levels).
    sp_factor (list): Sub-plot treatment names (e.g., Varieties).
    n_locs (int): Number of locations.
    n_blocks (int): Number of blocks (replicates) per location.
    
    Returns:
    dict: A dictionary containing the randomization DataFrame and the Matplotlib figure.
    """
    n_wp = len(wp_factor)
    n_sp = len(sp_factor)
    
    # ---------------------------------------------------------
    # 1. Randomization Engine
    # ---------------------------------------------------------
    plan_list = []
    
    for loc in range(1, n_locs + 1):
        for blk in range(1, n_blocks + 1):
            
            # Step 1: Randomize the Whole-plot factor within the block
            wp_rand = random.sample(list(wp_factor), n_wp)
            
            for w, current_wp in enumerate(wp_rand, start=1):
                
                # Step 2: Randomize the Sub-plot factor WITHIN the current Whole-plot
                sp_rand = random.sample(list(sp_factor), n_sp)
                
                # Assign spatial coordinates 
                temp_df = pd.DataFrame({
                    "Location": f"Location {loc}",
                    "Block": f"Block {blk}",
                    "Block_Num": blk,
                    "Whole_Plot_Unit": f"B{blk}-WP{w}",
                    "Whole_Plot_Trt": current_wp,
                    "Sub_Plot_Trt": sp_rand,
                    "Local_Row": w,                   # Row assignment based on WP index
                    "Local_Col": range(1, n_sp + 1)   # Column assignment based on SP index
                })
                
                plan_list.append(temp_df)
                
    final_plan = pd.concat(plan_list, ignore_index=True)
    
    # ---------------------------------------------------------
    # 2. Spatial Calculations
    # ---------------------------------------------------------
    
    # Stack blocks vertically
    final_plan['Global_Row'] = final_plan['Local_Row'] + (final_plan['Block_Num'] - 1) * n_wp
    final_plan['Global_Col'] = final_plan['Local_Col']
    
    # Bounding boxes for Whole-Plots (thicker gray lines)
    wp_boundaries = final_plan.groupby(['Location', 'Block', 'Whole_Plot_Unit']).agg(
        xmin=('Global_Col', lambda x: x.min() - 0.5),
        xmax=('Global_Col', lambda x: x.max() + 0.5),
        ymin=('Global_Row', lambda y: y.min() - 0.5),
        ymax=('Global_Row', lambda y: y.max() + 0.5)
    ).reset_index()
    
    # Bounding boxes for full Blocks (bold black lines)
    blk_boundaries = final_plan.groupby(['Location', 'Block']).agg(
        xmin=('Global_Col', lambda x: x.min() - 0.5),
        xmax=('Global_Col', lambda x: x.max() + 0.5),
        ymin=('Global_Row', lambda y: y.min() - 0.5),
        ymax=('Global_Row', lambda y: y.max() + 0.5)
    ).reset_index()
    
    # ---------------------------------------------------------
    # 3. Plotting
    # ---------------------------------------------------------
    
    fig, axes = plt.subplots(1, n_locs, figsize=(6 * n_locs, 8), squeeze=False)
    axes = axes.flatten()
    
    # Set up discrete color palette (matching Pastel1 from ggplot)
    unique_wp = final_plan['Whole_Plot_Trt'].unique()
    pastel_colors = plt.cm.Pastel1.colors
    color_map = {trt: pastel_colors[i % len(pastel_colors)] for i, trt in enumerate(unique_wp)}
    
    for i, loc in enumerate(final_plan['Location'].unique()):
        ax = axes[i]
        loc_data = final_plan[final_plan['Location'] == loc]
        
        # Draw individual Sub-plots
        for _, row in loc_data.iterrows():
            rect = patches.Rectangle(
                (row['Global_Col'] - 0.5, row['Global_Row'] - 0.5),
                1, 1, facecolor=color_map[row['Whole_Plot_Trt']],
                edgecolor='white', linewidth=0.5
            )
            ax.add_patch(rect)
            
            # Add Sub-plot treatment labels
            ax.text(row['Global_Col'], row['Global_Row'], row['Sub_Plot_Trt'],
                    ha='center', va='center', fontsize=9, fontweight='bold', color='black')
        
        # Draw Whole-plot boundaries
        loc_wp = wp_boundaries[wp_boundaries['Location'] == loc]
        for _, row in loc_wp.iterrows():
            rect = patches.Rectangle(
                (row['xmin'], row['ymin']), row['xmax'] - row['xmin'], row['ymax'] - row['ymin'],
                fill=False, edgecolor='dimgray', linewidth=1.5
            )
            ax.add_patch(rect)
            
        # Draw Block boundaries (Bold)
        loc_blk = blk_boundaries[blk_boundaries['Location'] == loc]
        for _, row in loc_blk.iterrows():
            rect = patches.Rectangle(
                (row['xmin'], row['ymin']), row['xmax'] - row['xmin'], row['ymax'] - row['ymin'],
                fill=False, edgecolor='black', linewidth=2.5
            )
            ax.add_patch(rect)
            
        # Facet Formatting
        ax.set_title(loc, fontweight='bold', fontsize=12)
        ax.set_xlim(0.5, n_sp + 0.5)
        ax.set_ylim(loc_data['Global_Row'].max() + 0.5, 0.5)  # Scale Y Reverse
        ax.set_xticks([])
        ax.set_yticks([])
        ax.set_xlabel("Sub-Plot (Column)")
        if i == 0:
            ax.set_ylabel("Whole-Plot (Row)")
        ax.set_aspect('equal')
        
    # Add Legend
    handles = [patches.Patch(color=color_map[trt], label=trt) for trt in unique_wp]
    fig.legend(handles=handles, title="Main Plot\nTreatment", loc='center right', bbox_to_anchor=(0.95, 0.5))
    
    # Titles and Subtitles
    plt.suptitle("Multi-Location Split-Plot Field Layout", fontsize=14, fontweight='bold')
    fig.text(0.5, 0.05, "Color = Whole-Plot Treatment. Text = Sub-Plot Treatment.\nBold Black = Block. Thin Gray = Whole-plot Strip.", 
             ha='center', fontsize=10, color='gray')
    
    plt.tight_layout(rect=[0, 0.08, 0.88, 0.95])
    plt.show()
    
    return {"Data": final_plan, "Plot": fig}

### ==========================================
### EXAMPLE USAGE
### ==========================================
##
##if __name__ == "__main__":
##    # Define 3 Main Plot treatments (e.g., Irrigation methods)
##    main_plots = ["Irrigated", "Dryland", "Deficit"]
##    
##    # Define 5 Sub-plot treatments (e.g., Varieties)
##    sub_plots = [f"Var-{i}" for i in range(1, 6)]
##    
##    # Generate the split-plot plan
##    split_trial = generate_stripplot_plan(
##        wp_factor=main_plots, 
##        sp_factor=sub_plots, 
##        n_locs=2, 
##        n_blocks=3
##    )
##    
##    # Check the generated dataset
##    print(split_trial["Data"].head(10))

File: core_logic/r_package/plantbreeding/R/generate_stripplot_plan.R
library(dplyr)
library(ggplot2)

#' Generate Strip-Plot Design Randomization Plan
#'
#' @param wp_factor Vector of Whole-plot treatment names (e.g., Irrigation levels).
#' @param sp_factor Vector of Sub-plot treatment names (e.g., Varieties).
#' @param n_locs Number of locations.
#' @param n_blocks Number of blocks (replicates) per location.
#'
#' @return A list containing the randomization data frame and the ggplot object.
generate_stripplot_plan <- function(wp_factor, sp_factor, n_locs = 2, n_blocks = 3) {
  
  n_wp <- length(wp_factor)
  n_sp <- length(sp_factor)
  
  # ---------------------------------------------------------
  # 1. Randomization Engine
  # ---------------------------------------------------------
  
  plan_list <- list()
  counter <- 1
  
  for (loc in 1:n_locs) {
    for (blk in 1:n_blocks) {
      
      # Step 1: Randomize the Whole-plot factor within the block
      wp_rand <- sample(wp_factor)
      
      for (w in 1:n_wp) {
        current_wp <- wp_rand[w]
        
        # Step 2: Randomize the Sub-plot factor WITHIN the current Whole-plot
        sp_rand <- sample(sp_factor)
        
        # Assign spatial coordinates 
        # (Whole-plots = Rows within block, Sub-plots = Columns within Whole-plot)
        temp_df <- data.frame(
          Location = paste("Location", loc),
          Block = paste("Block", blk),
          Block_Num = blk,
          Whole_Plot_Unit = paste0("B", blk, "-WP", w),
          Whole_Plot_Trt = current_wp,
          Sub_Plot_Trt = sp_rand,
          Local_Row = w,               # Row assignment based on WP index
          Local_Col = 1:n_sp           # Column assignment based on SP index
        )
        
        plan_list[[counter]] <- temp_df
        counter <- counter + 1
      }
    }
  }
  
  final_plan <- bind_rows(plan_list)
  
  # ---------------------------------------------------------
  # 2. Spatial Calculations
  # ---------------------------------------------------------
  
  # Stack blocks vertically
  final_plan <- final_plan %>%
    group_by(Location) %>%
    mutate(
      Global_Row = Local_Row + (Block_Num - 1) * n_wp,
      Global_Col = Local_Col
    ) %>%
    ungroup()
  
  # Bounding boxes for Whole-Plots (thicker gray lines)
  wp_boundaries <- final_plan %>%
    group_by(Location, Block, Whole_Plot_Unit) %>%
    summarize(
      xmin = min(Global_Col) - 0.5,
      xmax = max(Global_Col) + 0.5,
      ymin = min(Global_Row) - 0.5,
      ymax = max(Global_Row) + 0.5,
      .groups = "drop"
    )
  
  # Bounding boxes for full Blocks (bold black lines)
  blk_boundaries <- final_plan %>%
    group_by(Location, Block) %>%
    summarize(
      xmin = min(Global_Col) - 0.5,
      xmax = max(Global_Col) + 0.5,
      ymin = min(Global_Row) - 0.5,
      ymax = max(Global_Row) + 0.5,
      .groups = "drop"
    )
  
  # ---------------------------------------------------------
  # 3. Plotting
  # ---------------------------------------------------------
  
  field_plot <- ggplot() +
    # Draw individual Sub-plots
    geom_tile(data = final_plan, 
              aes(x = Global_Col, y = Global_Row, fill = Whole_Plot_Trt), 
              color = "white", linewidth = 0.5) +
    
    # Draw Whole-plot boundaries
    geom_rect(data = wp_boundaries, 
              aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), 
              fill = NA, color = "gray20", linewidth = 0.8) +
    
    # Draw Block boundaries (Bold)
    geom_rect(data = blk_boundaries, 
              aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), 
              fill = NA, color = "black", linewidth = 1.5) +
    
    # Add Sub-plot treatment labels
    geom_text(data = final_plan, 
              aes(x = Global_Col, y = Global_Row, label = Sub_Plot_Trt), 
              size = 3.5, fontface = "bold", color = "black") +
    
    scale_y_reverse() + 
    scale_x_continuous() +
    facet_wrap(~ Location, ncol = 2, scales = "free") + 
    
    # Using a discrete color palette for the Whole-plot treatments
    scale_fill_brewer(palette = "Pastel1") + 
    
    labs(title = "Multi-Location Split-Plot Field Layout",
         subtitle = "Color = Whole-Plot Treatment. Text = Sub-Plot Treatment.\nBold Black = Block. Thin Gray = Whole-plot Strip.",
         x = "Sub-Plot (Column)",
         y = "Whole-Plot (Row)",
         fill = "Main Plot\nTreatment") +
    theme_minimal() +
    theme(
      panel.grid = element_blank(),
      strip.text = element_text(size = 12, face = "bold"),
      axis.text = element_blank() 
    )
  
  print(field_plot)
  return(list(Data = final_plan, Plot = field_plot))
}

# # ==========================================
# # EXAMPLE USAGE
# # ==========================================
# 
# # Define 3 Main Plot treatments (e.g., Irrigation methods)
# main_plots <- c("Irrigated", "Dryland", "Deficit")
# 
# # Define 5 Sub-plot treatments (e.g., Varieties)
# sub_plots <- paste0("Var-", 1:5)
# 
# # Generate the split-plot plan
# # This will result in blocks of 3 rows (Main plots) and 5 columns (Sub-plots)
# split_trial <- generate_stripplot_plan(
#   wp_factor = main_plots, 
#   sp_factor = sub_plots, 
#   n_locs = 2,             # 2 environments
#   n_blocks = 3            # 3 replicates per environment
# )
# 
# # Check the generated dataset
# head(split_trial$Data, 10)